internal/resources/gsi.go

```diff
@@ -0,0 +1,509 @@
+			tflog.Info(ctx, "resource doesn't exist in remote server removing resource from state file")
+			resp.State.RemoveResource(ctx)
+			return
+		}
+
+		resp.Diagnostics.AddError(
+			"Error reading query index",
+			"Could not read query index "+state.IndexName.ValueString()+": "+errString,
+		)
+		return
+	}
+
+	if !state.OrganizationId.IsNull() {
+		// when reading an index, only update number of replicas.
+		state.With.NumReplica = types.Int64Value(int64(index.NumReplica))
+
+	} else {
+		// when importing index, set all attributes.
+		state.OrganizationId = types.StringValue(organizationID)
+		state.ProjectId = types.StringValue(projectID)
+		state.ClusterId = types.StringValue(clusterID)
+		state.BucketName = types.StringValue(bucketName)
+		state.ScopeName = types.StringValue(scopeName)
+		state.CollectionName = types.StringValue(collectionName)
+		state.IsPrimary = types.BoolValue(index.IsPrimary)
+		state.IndexName = types.StringValue(indexName)
+
+		var keys []attr.Value
+		for _, key := range index.SecExprs {
+			keys = append(keys, types.StringValue(key))
+		}
+
+		keyList, newDiags := types.ListValue(types.StringType, keys)
+		if newDiags.HasError() {
+			resp.Diagnostics.AddError(
+				"Error converting index keys to set type",
+				"Could not convert index keys to set type for index "+state.IndexName.ValueString()+": "+err.Error(),
+			)
+			return
+		}
+
+		state.IndexKeys = keyList
+		// TODO:  set partition by.
+		state.Where = types.StringValue(index.Where)
+		if state.With != nil {
+			state.With.NumReplica = types.Int64Value(int64(index.NumReplica))
+			state.With.NumPartition = types.Int64Value(int64(index.NumPartition))
+		}
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (g *GSI) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	ddl := fmt.Sprintf(
+		"ALTER INDEX `%s` ON `%s`.`%s`.`%s` WITH { \"action\": \"replica_count\", \"num_replica\" : %d }",
+		plan.IndexName.ValueString(),
+		plan.BucketName.ValueString(),
+		plan.ScopeName.ValueString(),
+		plan.CollectionName.ValueString(),
+		plan.With.NumReplica.ValueInt64(),
+	)
+
+	if err := g.executeGsiDdl(ctx, &plan, ddl); err != nil {
+		resp.Diagnostics.AddError(
+			"An error occurred while executing index DDL",
+			"Error during index DDL execution: "+err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (g *GSI) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var state providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !state.BuildIndexes.IsNull() {
+		return
+	}
```

Comment:
Same as above

Can you please add a GoDoc comment explaining why we are returning here?

Reply:
Addressed in 5c6efe534d7761f3957ee6e5d1301148aba9aa2c